pkg/blockbuilder/scheduler/scheduler.go

```diff
@@ -121,116 +114,597 @@ func (s *BlockBuilderScheduler) running(ctx context.Context) error {
+
+				consumeOffset = committedOff.At
+			} else {
+				// Nothing committed for this partition. Rewind to fallback offset instead.
+				o, ok := fallbackOffsets.Lookup(t, partition)
+				if !ok {
+					return nil, fmt.Errorf("partition %d not found in fallback offsets for topic %s", partition, t)
+				}
+
+				level.Debug(s.logger).Log("msg", "no commit; falling back to max of startOffset and fallbackOffset",
+					"topic", t, "partition", partition, "startOffset", startOffset.At, "fallbackOffset", o.Offset)
+
+				consumeOffset = max(startOffset.At, o.Offset)
+			}
+
+			end, ok := endOffsets.Lookup(t, partition)
+			if !ok {
+				return nil, fmt.Errorf("partition %d not found in end offsets for topic %s", partition, t)
+			}
+
+			level.Debug(s.logger).Log("msg", "consumptionOffsets", "topic", t, "partition", partition, "start", startOffset.At, "end", end.Offset, "consumeOffset", consumeOffset)
+
+			s.metrics.partitionStartOffset.WithLabelValues(partStr).Set(float64(startOffset.At))
+			s.metrics.partitionEndOffset.WithLabelValues(partStr).Set(float64(end.Offset))
+
+			offs = append(offs, partitionOffsets{
+				topic:     t,
+				partition: partition,
+				start:     startOffset.At,
+				resume:    consumeOffset,
+				end:       end.Offset,
+			})
+		}
+	}
+
+	return offs, nil
+}
+
+type offsetStore interface {
+	offsetAfterTime(context.Context, string, int32, time.Time) (int64, time.Time, error)
+}
+
+type offsetFinder struct {
+	offsets     map[time.Time]kadm.ListedOffsets
+	adminClient *kadm.Client
+	logger      log.Logger
+}
+
+func newOffsetFinder(adminClient *kadm.Client, logger log.Logger) *offsetFinder {
+	return &offsetFinder{
+		offsets:     make(map[time.Time]kadm.ListedOffsets),
+		adminClient: adminClient,
+		logger:      logger,
+	}
 }
 
-func (s *BlockBuilderScheduler) fetchLag(ctx context.Context) (kadm.GroupLag, error) {
+// offsetAfterTime is a cached version of adminClient.ListOffsetsAfterMilli that
+// makes use of the fact that we want to ask about the same times for all partitions.
+func (o *offsetFinder) offsetAfterTime(ctx context.Context, topic string, partition int32, t time.Time) (int64, time.Time, error) {
+	offs, ok := o.offsets[t]
+	if !ok {
+		var err error
+		offs, err = o.adminClient.ListOffsetsAfterMilli(ctx, t.UnixMilli(), topic)
+		if err != nil {
+			return 0, time.Time{}, err
+		}
+		if offs.Error() != nil {
+			return 0, time.Time{}, offs.Error()
+		}
+
+		o.offsets[t] = offs
+	}
+
+	po, ok := offs.Lookup(topic, partition)
+	if !ok {
+		return 0, time.Time{}, fmt.Errorf("failed to get offset for partition %d at time %s: not present", partition, t)
+	}
+	if po.Err != nil {
+		return 0, time.Time{}, fmt.Errorf("failed to get offset for partition %d at time %s: %w", partition, t, po.Err)
+	}
+
+	return po.Offset, time.UnixMilli(po.Timestamp), nil
+}
+
+var _ offsetStore = (*offsetFinder)(nil)
+
+// fetchCommittedOffsets fetches the committed offsets for the given consumer group.
```

Comment:
(uber-nit)

```suggestion
// fetchCommittedOffsets fetches the committed offsets for the scheduler's consumer group.
```

The caller of the `fetchCommittedOffsets` method doesn't influence on the consumer group used.